crates/wysiwyg/src/composer_model/code_block.rs

```diff
@@ -0,0 +1,523 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::dom::action_list::DomActionList;
+use crate::dom::nodes::dom_node::DomNodeKind;
+use crate::dom::nodes::dom_node::DomNodeKind::*;
+use crate::dom::nodes::{ContainerNodeKind, DomNode};
+use crate::dom::unicode_string::UnicodeStr;
+use crate::dom::{DomHandle, DomLocation};
+use crate::{
+    ComposerAction, ComposerModel, ComposerUpdate, ToHtml, UnicodeString,
+};
+use std::cmp::min;
+use std::collections::HashSet;
+use std::ops::{Add, AddAssign};
+
+impl<S> ComposerModel<S>
+where
+    S: UnicodeString,
+{
+    pub fn code_block(&mut self) -> ComposerUpdate<S> {
+        if self.action_is_reversed(ComposerAction::CodeBlock) {
+            // TODO: add code block removal if selection is inside the code block, otherwise extend it
+            self.remove_code_block()
+        } else {
+            self.add_code_block()
+        }
+    }
+
+    fn add_code_block(&mut self) -> ComposerUpdate<S> {
+        let (s, e) = self.safe_selection();
+        let Some((parent_handle, idx_start, idx_end)) = self.find_nodes_to_wrap_in_block(s, e) else {
+            return ComposerUpdate::keep();
+        };
+        let mut leaves_to_add: Vec<DomNode<S>> = Vec::new();
+        let mut nodes_visited = HashSet::new();
+        let start_handle = parent_handle.child_handle(idx_start);
+        let up_to_handle = parent_handle.child_handle(idx_end + 1);
+        let iter = self.state.dom.iter_from_handle(&start_handle);
+        for node in iter {
+            if node.handle() >= up_to_handle {
+                break;
+            }
+            match node.kind() {
+                Text | Formatting(_) | Link => {
+                    nodes_visited.insert(node.handle());
+                    if nodes_visited.contains(&node.handle().parent_handle()) {
+                        continue;
+                    }
+                    leaves_to_add.push(node.clone());
+                }
+                LineBreak => {
+                    leaves_to_add.push(DomNode::new_text("\n".into()));
+                }
+                List | ListItem => {
+                    let mut needs_to_add_line_break =
+                        node.handle().index_in_parent() > 0;
```

Comment:
Is it right that the code block is coupled so tightly to lists?

I'm wondering if we can reduce code block's knowledge of the other node types so it doesn't need to be updated each time we add or change a node type? If for example, all nodes types implement a `to_code()` method, then code block can simply call `node.to_code()` on any compatible node type and concentrate on the wrapping / unwrapping the result.

Reply:
I added a `ComposerModel.format_node_for_code_block` function, but we still need to check the kind of the DomNode to understand where we need to insert new line breaks.